### ISSUE TITLE
Support .NET Platform Standard 1.6 (inc. .NET Core 1.0)

### ISSUE THREAD

**Issue by khellang** [FakeItEasy/FakeItEasy#531]
I figured it's time to start this discussion :smile:

Here are some threads for reference:

https://github.com/castleproject/Core/issues/90
https://github.com/nsubstitute/NSubstitute/issues/192
https://github.com/Moq/moq4/issues/168

There's a PR pending for .NET Core support in Castle.Core at https://github.com/castleproject/Core/pull/92

**Comment by adamralph** [comment]
As soon as Castle.Core support the dotnet TFN I'm happy to do this.

**Comment by jeremymeng** [comment]
Would it be okay to start the work based on https://github.com/castleproject/Core/tree/netcore? Most of the big changes are already in. I am willing to give it a first stab.

**Comment by adamralph** [comment]
@jeremymeng I think would be fine to grab some custom binary to reference for now, so long as your are prepared to rebase/squash your commits back to something clean (or have someone else do that) when Castle for netcore (or whatever other name is goes by) is on NuGet.

**Comment by jchannon** [comment]
Where are we at with this?

**Comment by thecodejunkie** [comment]
We have a green build against rc2 now (including tests, i.e all tests that do not require FakeItEasy) https://github.com/NancyFx/Nancy/pull/2418 and would be happy to help transfer knowledge on rc2 changes to help this move along

**Comment by blairconrad** [comment]
Hey, @jchannon and @thecodejunkie. As you can see, @jeremymeng had kindly been doing some work on this for us, but the core team had all been focusing on FIE 2.0, which I think we're nearly ready to release the second (and final, one hopes) RC for.

We'd be grateful for any insight you have to offer, but I'm not going to be available for the next two weeks, so can't personally do anything. I'll leave it up to @thomaslevesque and @adamralph to coordinate with you.

**Comment by jeremymeng** [comment]
@jchannon @thecodejunkie 

There are a couple of work items now
- [ ] Replacement of binary serialization used in `FileStorage`.  So far I haven't had success in using a Json or Xml serializer.
- [ ] Seemly incorrect usage of my custom AssemblyLoadContext.  No tests are hitting the code path where I made changes.  I am not sure how to test it yet.
- [ ] Move to RC2
- [ ] integration into build system, etc.

Any helps are welcome :smile:

**Comment by thomaslevesque** [comment]
@jeremymeng, thanks for all your work on this task!

> Any helps are welcome :smile:

Right now I'm focusing on FakeItEasy 2.0.0-rc2, but once it's released, I'll be glad to give you a hand. I know almost nothing about .NET Core, though...

**Comment by adamralph** [comment]
Given the recent sweeping changes and volatile nature of .NET Core, I'm not sure it's worth investing any time in this in the near future. I'm leaning towards closing for now, and opening a new issue to support the relevant [.NET Standard](https://gist.github.com/davidfowl/8939f305567e1755412d6dc0b8baf1b7) after RTM.

We appreciate the efforts of @jeremymeng and he is, of course, welcome to carry on with his work on  https://github.com/FakeItEasy/FakeItEasy/pull/617 but right now, if we have no intention of supporting .NET core until some later release, I believe this issue is just noise in our backlog.

**Comment by thomaslevesque** [comment]
@adamralph, RC2 is scheduled to be release about now (mid-May), RTM is scheduled for the end of June. Sounds like "near future" to me :wink:. This issue is still relevant, so I think we should keep it open.

**Comment by adamralph** [comment]
OK, np. Let's leave it open.

**Comment by henningst** [comment]
RTM shipped as planned on June 27th. Do you have any ETA on adding .NET Core support? Moq has published an alpha supporting .NET Core.

**Comment by blairconrad** [comment]
Thanks for the interest, @henningst. We don't have an ETA as such (given that we all work on FakeItEasy if and when we have spare time to so, it's hard to guess at dates), but rest assured that we do want to support .NET Core. The same chap (@jeremymeng) that's been working on adding .NET support to Castle.Core (which, like Moq, has alphas that currently support the prerelease .NET Core) has been making great strides in helping FakeItEasy support .NET Core. As we're able we'll look at how best to integrate his work.

**Comment by jeremymeng** [comment]
People who are interested can download private nupkgs here https://ci.appveyor.com/project/jeremymeng/fakeiteasy/build/artifacts

**Comment by adamralph** [comment]
Thanks @jeremymeng.

@henningst if you like, you can use @jeremymeng's app veyor build as your "alpha" channel for now.

**Comment by henningst** [comment]
Thanks for the tip! I'll give that a try :)

**Comment by blairconrad** [comment]
For those requiring an alpha NuGet feed, @jeremymeng has [made available](https://twitter.com/JeremyYMeng/status/753739438614720516)

https://ci.appveyor.com/nuget/fakeiteasy-jeremymeng

He'd had a [caveat](https://twitter.com/JeremyYMeng/status/753740007483936769)

> You'd want to avoid v2.3.0-coreclr-alpha21 as it is not the latest (minor version shouldn't have been bumped)

But I think it no longer applies, since we've released 2.2.0.  2.3.0-coreclr-alpha30 is the state of the art, as I write this.

**Comment by adamralph** [comment]
Bumped to P1 since it seems the most pressing issue right now.

**Comment by adamralph** [comment]
Congratulations on getting in https://github.com/FakeItEasy/FakeItEasy/pull/617. Great work all round. :clap: 

I have a few questions/concerns about the final commit (I apologise for raising these now instead of when I should have, which was during the review phase of the PR):

* when I build the project in VS, I get:

> Warning	NU1012	Dependency conflict. xunit.runner.utility 2.2.0-beta2-build3300 expected xunit.abstractions >= 2.0.1-rc2 but received 2.0.0	FakeItEasy.Tests.dotnet	C:\code\FakeItEasy\FakeItEasy\tests\FakeItEasy.Tests.dotnet\project.json	27	

* Why do we have a separate solution for the `dotnet` projects? If it has to remain, should it be named `FakeItEasy-dotnet.sln` for consistency with the projects?
* Why has [this](https://github.com/FakeItEasy/FakeItEasy/blob/46becf4a6cc8d72ddefc8ca024167d29979efdde/FakeItEasy.sln#L38-L40) project section been added?
* `global.json` seems wrong. It includes `test` instead of `tests` and omits `samples`.
* What is the purpose of the `wrap` folder? @blairconrad has shed [some light](https://github.com/FakeItEasy/FakeItEasy/commit/46becf4a6cc8d72ddefc8ca024167d29979efdde#commitcomment-18460354) but I still don't fully understand. Could someone please explain?
* Why do we have to pass the packages folder into the msbuild command? Are we using a non-standard package folder location?
* Is `dotnet` the best suffix/name to distinguish the new bits? Since the aim is to support the `netstandard1.6` TFN I would have thought `netstandard` would be more appropriate. (Or `netstd`.)
* The dotnet project is missing:
  * `CodeAnalysisRuleSet`
  * `StyleCop.MSBuild`
  * `DebugSymbols` in `Release`
  * `TreatWarningsAsErrors` `true` in `Release`
* We are missing an approval test project for the dotnet project
* Why are we using csproj for FakeItEasy.dotnet and xproj for FakeItEasy.Tests.dotnet?


And a few nitpicks:

* `.nuget.targets` in `.gitignore` seems redundant. I couldn't find a way to get such a file to appear.
* The `.vs` folder has been ignored in a manner inconsistent with the other entries. I think it should be just `.vs` and appear at the top of the file. (`packages` is also in the "wrong" place but that was added a long time ago.)
* The `wrap` folder should be added to the solution(s), just like all the other folders.
* `global.json` should be added to "Solution Items" in FakeItEasy.sln, even if it's not relevant in that context (IMHO all files in the root folder should be added). (`build.cmd` is also missing in both solutions but that was added to the folder a while ago.)
* Is there a reason the FakeItEasy project GUID changed?
* There are few places where things have been moved within files unnecessarily. E.g. [here](https://github.com/FakeItEasy/FakeItEasy/commit/46becf4a6cc8d72ddefc8ca024167d29979efdde#diff-67904ac000794fc8a1575286c6821ab9R122) and [here](https://github.com/FakeItEasy/FakeItEasy/commit/46becf4a6cc8d72ddefc8ca024167d29979efdde#diff-3c1006d0d14ca73cb2625e596e45b2e3R65) (and many more).
* The dotnet project is producing a docs file in Debug. It should only do this in Release.
* The docs file for dotnet project has the extension capitalised as "XML" instead of "xml".
* There seems to have been lots of unnecessary churn inside FakeItEasy.csproj. Most of it seems to be *good* churn, like removing `<SubType>Code</SubType>` but I would have liked to see this in a separate refactoring commit.
* Whilst I :heart: the changed from `Bin` to `bin`, it should also have been a separate refactoring commit.

**Comment by adamralph** [comment]
One more question:

* If we were to [remove self initialised fakes](https://github.com/FakeItEasy/FakeItEasy/issues/798) would that leave the net40 and netstandard1.6 libs functionally identical? In that case, I guess we'd be keeping the net40 lib around to support .NET 4.0 - 4.6.2, since netstandard1.6 only supports (the still be released?) .NET 4.6.3?

Another nitpick

* The [addition of a dependency group](https://github.com/FakeItEasy/FakeItEasy/commit/46becf4a6cc8d72ddefc8ca024167d29979efdde#diff-12c9a082caec06970f2411825bd256f2R16) in the nuspec seems redundant.

Given I've listed so many questions/concerns, I've added checkboxes next to them so I can track which have been answered/resolved. :grin:

**Comment by blairconrad** [comment]
Hi, @adamralph. Let's see what we can do about those comments. I don't have all the answers, and may be making bad guesses about some.

>  Why do we have a separate solution for the dotnet projects?

It allows people who don't have .NET core to open the solution and build it. I think we hadn't yet committed to requiring the tooling.

> If it has to remain, should it be named FakeItEasy-dotnet.sln for consistency with the projects?

No objection.

> What is the purpose of the wrap folder? @blairconrad has shed some light but I still don't fully understand. Could someone please explain?

> If we were to remove self initialised fakes would that leave the net40 and netstandard1.6 libs functionally identical?

I believe so. Of course, the exceptions and other types won't be binary serializable in the netstandard1.6 version, but as functionally identical as could be.

>  Is there a reason the FakeItEasy project GUID changed?

>  There are few places where things have been moved within files unnecessarily. E.g. here and here (and many more).

It's true. Do you prefer to restore the old positions? If the new arrangement isn't harmful, I think an additional commit would just be more noise at this point.

> If we were to remove self initialised fakes would that leave the net40 and netstandard1.6 libs functionally identical?

I think so, within the confines of the netstandard framework (no binary serialization for other types, like the fakes themselves, and our exceptions).

**Comment by adamralph** [comment]
@blairconrad thanks for the answers.

> the exceptions and other types won't be binary serializable in the netstandard1.6 version, but as functionally identical as could be.

(I think you answered this question twice.) By the time we decide to drop support for .NET 4.0 - 4.6.2 (probably not for a long time yet), binary serialization will probably be back in .NET Core. At that point we can switch to a single netstandard DLL.

> If the new arrangement isn't harmful, I think an additional commit would just be more noise at this point.

I agree. Let's leave it. I'd like to us to try and avoid this in the future though.

**Comment by blairconrad** [comment]
> > If the new arrangement isn't harmful, I think an additional commit would just be more noise at this point.

> I agree. Let's leave it. I'd like to us to try and avoid this in the future though.

Me too. I'd hazard that everyone feels the same.
I think we've all learned a lesson about getting in front of large changes such as this _before_ they become large changes such as this.

**Comment by blairconrad** [comment]
Revisiting

> What is the purpose of the wrap folder? 

I dug up @jeremymeng's [comment](https://github.com/FakeItEasy/FakeItEasy/pull/617#discussion_r69770949):

> Currently the PCL output assembly cannot be directly consumed by the project.json/xproj based projects. They can only reference project.json based projects or nuget packages. This wraps the output so it can be used by the .NET Core unit test project.

Of course, that may bring us back to 

> Why are we using csproj for FakeItEasy.dotnet

And I can't help with that.

**Comment by blairconrad** [comment]
> The addition of a dependency group in the nuspec seems redundant.

Yup. But harmless. Easy to remove.

**Comment by blairconrad** [comment]
> Is there a reason the FakeItEasy project GUID changed?

I think it was the way @jeremymeng had first introduced and cleaned up the .NET Core and 4.5 projects.
First, he'd moved FakeItEasy to FakeItEasy.Net40 and created a 4.5-targeting FakeItEasy. When he [removed the 4.5 project](a63dc962183238189a723c1cd447204fbbef561c), it looks like he changed the version that FakeItEasy targeted and deleted FakeItEasy.Net40.

**Comment by blairconrad** [comment]
> when I build the project in VS, I get:
> > Warning NU1012 Dependency conflict. xunit.runner.utility 2.2.0-beta2-build3300 expected xunit.abstractions >= 2.0.1-rc2 but received 2.0.0 FakeItEasy.Tests.dotnet C:\code\FakeItEasy\FakeItEasy\tests\FakeItEasy.Tests.dotnet\project.json 27

Because the project relies on two packages:

1. [dotnet-test-xunit 2.2.0-preview2-build1029](https://www.nuget.org/packages/dotnet-test-xunit/2.2.0-preview2-build1029), which relies on 
  * xunit.runner.reporters (>= 2.2.0-beta2-build3300) (2.2.0-beta2-build3300 is the highest), which relies on
    * xunit.runner.utility (= 2.2.0-beta2-build3300) which relies on 
      * xunit.abstractions (>= 2.0.1-rc2)
1. xunit.abstractions 2.0.0

the latter package had been 2.0.1-rc2 until [it was reverted to 2.0](9dd41ceae60d5c9473f8d1769e4f4d82eb987941).

Changing the dependencies to

```json
"dependencies": {
  "xunit.abstractions": "2.0.1-rc2",
  "xunit.extensibility.core": "2.2.0-beta2-build3300",
  "xunit.extensibility.execution": "2.2.0-beta2-build3300"
},
```

Resolves the conflict but when I `dotnet run`, I see:

```
FakeItEasy.Tests.dotnet [coreclr-mostly +0 ~1 -0]> dotnet run -c Release
Project FakeItEasy.Tests (.NETCoreApp,Version=v1.0) will be compiled because expected inputs are missing
Compiling FakeItEasy.Tests for .NETCoreApp,Version=v1.0

Compilation succeeded.
    0 Warning(s)
    0 Error(s)

Time elapsed 00:00:02.7637414

Unhandled Exception: System.MissingMethodException: Entry point not found in assembly 'FakeItEasy.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=eff28e2146d5fd2c'.
```

That's as far as I've gotten.

**Comment by blairconrad** [comment]
... I was on the coreclr branch. On master, the tests run.

```
FakeItEasy.Tests.dotnet [master +0 ~1 -0]> dotnet test -c Release
Project FakeItEasy.Tests (.NETCoreApp,Version=v1.0) will be compiled because expected inputs are missing
Compiling FakeItEasy.Tests for .NETCoreApp,Version=v1.0

Compilation succeeded.
    0 Warning(s)
    0 Error(s)

Time elapsed 00:00:02.8123046

xUnit.net .NET CLI test runner (64-bit .NET Core win10-x64)
  Discovering: FakeItEasy.Tests
  Discovered:  FakeItEasy.Tests
  Starting:    FakeItEasy.Tests
  Finished:    FakeItEasy.Tests
=== TEST EXECUTION SUMMARY ===
   FakeItEasy.Tests  Total: 1016, Errors: 0, Failed: 0, Skipped: 0, Time: 5.168s
SUMMARY: Total: 1 targets, Passed: 1, Failed: 0.
```

Submitting FakeItEasy/FakeItEasy#811. Unfortunately, the CI server won't run the code.

**Comment by blairconrad** [comment]
>  The dotnet project is producing a docs file in Debug. It should only do this in Release.
>  The docs file for dotnet project has the extension capitalised as "XML" instead of "xml".

#812

**Comment by blairconrad** [comment]
> The .vs folder has been ignored in a manner inconsistent with the other entries. I think it should be just .vs and appear at the top of the file. (packages is also in the "wrong" place but that was added a long time ago.)

#813

**Comment by blairconrad** [comment]
Before attempting to address some of the other concerns, I think we should make a decision about forcing contributors to have .NET Core tooling.

Shall we add the .dotnet (or however they might be named) projects to the `FakeItEasy.sln` and remove `FakeItEasy-NetCore.sln`?

This will affect how we clean up the sln and our rakefile work.

**Comment by adamralph** [comment]
Thanks for the PR's @blairconrad. Those check boxes are seeing some action!

I think it's OK to leave the solutions in place for now. I think it's too early to force the half baked tooling on our contributors.

**Comment by blairconrad** [comment]
Thanks for the merges!

Leaving the sln suits me fine.

In that case, I propose adding targets to the `rakefile` so we can build our netstandard projects (and test them and so on).

The best option I can think of is parallel targets, distinguished by affix, such as

* build_dotnet
* unit_dotnet
* default_dotnet (or just dotnet)

Of course we can pick a different affix (and standardize on the project naming while we're at it, as @adamralph suggested earlier). But this could be adjusted after the fact. My preference would be to hook up easy building/test running and _then_ muck with the names.

After we have rake targets, I think we should add them to the CI build. We could make them optional, like `approve`, so as not to scare off the contributors who don't have the new tooling, or we could make them required. For me, they feel optional at this point, but I won't argue if you guys prefer required.

Does this make sense?

**Comment by adamralph** [comment]
:+1: makes sense to me. I'm fine with sticking to `dotnet` for now but I would like to globally change to netstd later. `dotnet` is a [deprecated monkier](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md#deprecated-monikers).

I would go with `dotnet` as the parallel task to `default`.

I think `dotnet` should be a superset of `default`. I.e. the tasks it runs should perform all actions for both the `net40` and `netstandard` assemblies. Then those who have the tooling only ever have to run `bundle exec rake dotnet` (or `build.cmd dotnet`) and the CI server can do the same.

Are you sure you want to do this before changing the `dotnet` moniker? If we change that now there is less to change.

**Comment by blairconrad** [comment]
Okay. I nearly have a thing. I'll submit a WIP PR soon.

> Are you sure you want to do this before changing the `dotnet` moniker? If we change that now there is less to change.

I'm not that fussy. I think updating the rakefile will not be such a big deal and figure that it'll take us a week to agree on a moniker. You've already used `netstd` and `netstandard` in your previous comment. :wink:

I'll send up my PR for parallel look-atting, and if we decide on a name in the meantime, I'll update. I don't mind.

**Comment by adamralph** [comment]
OK, no problem. Let's rename later. You're probably right about us taking a week to agree on the moniker :stuck_out_tongue:

**Comment by blairconrad** [comment]
For your dissection, #814

**Comment by adamralph** [comment]
On second thoughts, should this be in 3.0? It doesn't have to be. E.g. we could push out 3.0 with the breaking changes we want to do now, and push this out later. However, given that Castle.Core with netstd support isn't out of beta yet, I'm not sure it's worth taking the risk that we *may* have to break something else before releasing RTM netstd support.

**Comment by adamralph** [comment]
After changing the title, this question arose in my mind:

* Is there a good reason we are supporting only netstd 1.6 and above? The Castle.Core beta support netstd 1.3, which is compatible with more platforms.

(Added to list of questions above.)

**Comment by thomaslevesque** [comment]
> Is there a good reason we are supporting only netstd 1.6 and above? The Castle.Core beta supports netstd 1.3, which is compatible with more platforms.

I think it's because of this dependency, which only supports 1.6:

https://www.nuget.org/packages/Microsoft.Extensions.DependencyModel/

I'm not sure why we need it, though... it says "Abstractions for reading `.deps` files.", no idea what that means.

**Comment by adamralph** [comment]
Thanks for finding that @thomaslevesque.

@jeremymeng can you shed any light on why we need to depend on https://www.nuget.org/packages/Microsoft.Extensions.DependencyModel/ ?

**Comment by blairconrad** [comment]
> Why do we not have specs and integration tests for dotnet?

Mostly because we've not got 'round to them yet. I talk about the integration tests in https://github.com/FakeItEasy/FakeItEasy/pull/815#issuecomment-236899715

But the specs are up for grabs. Then again, given my [problem compiling the dotnet unit tests](https://github.com/FakeItEasy/FakeItEasy/pull/815#discussion_r73132820), maybe we should see how to resolve that before involving more projects.

**Comment by blairconrad** [comment]
> @jeremymeng can you shed any light on why we need to depend on https://www.nuget.org/packages/Microsoft.Extensions.DependencyModel/ ?

It's used in the bootstrapper and other extension point scanning. For figuring out which assemblies reference which other ones so we can load the right types. See [TypeCatalogue](https://github.com/FakeItEasy/FakeItEasy/blob/46becf4a6cc8d72ddefc8ca024167d29979efdde/src/FakeItEasy/Core/TypeCatalogue.cs) for starters.

**Comment by blairconrad** [comment]
> On second thoughts, should this be in 3.0? It doesn't have to be. E.g. we could push out 3.0 with the breaking changes we want to do now, and push this out later. However, given that Castle.Core with netstd support isn't out of beta yet, I'm not sure it's worth taking the risk that we may have to break something else before releasing RTM netstd support.

I am not sure which way you're arguing now.

I'm worried about committing to a 3.0 that includes .NET Standard support because we don't know how long it'll be 'til we get that.

I wouldn't mind delaying the .NET Standard-supporting release. If you prefer to hold off on releasing another major build until we can do that, that's fine with me. Just let's not start throwing breaking changes (without toggles?) into the code in the meantime?

**Comment by jeremymeng** [comment]
> the latter package had been 2.0.1-rc2 until it was reverted to 2.0.

(I am on vacation so please forgive my delayed response.) 

I attempted to fix the warning but haven't succeeded.  There are recommended xunit versions at https://xunit.github.io/docs/getting-started-dotnet-core.html I wanted to try out but haven't done so.

**Comment by jeremymeng** [comment]
> I'm not sure why we need it, though... it says "Abstractions for reading .deps files.", no idea what that means.

I'd thought I have commented somewhere but I might have forgotten to click the Comment button.  .NET Core build generates a `<assembly>.dpes.json` file along side of the output assembly.  I believe `DependencyModel` is providing compilation/dependency information within this file at the run time.

**Comment by jeremymeng** [comment]
I also heard that this code (`DependencyModel`) might be going away in the future.  I don't know more details yet.

**Comment by blairconrad** [comment]
Thanks, @jeremymeng. No need to apologize fro delays. You're under no obligation!
In theory StructureMap is supporting _some kind_ of scanning while relying on NetStandard 1.3 but I've not had time to investigate. It may not be the same kind of thing.

**Comment by blairconrad** [comment]
Oh, and enjoy your vacation, @jeremymeng!

**Comment by jeremymeng** [comment]
> Why has this project section been added?

Because the example project actually depends on the unit test project.  I unloaded the NET45 unit test project so that I only have .NET Core tests showing in the VS test explorer.  I added this project dependency so that when the NET45 unit test project is not built, VS doesn't attempt to build the example project.

**Comment by jeremymeng** [comment]
> global.json seems wrong. It includes test instead of tests and omits samples.

Yes, it should be `tests` or rather it should be removed.  `dotnet restore` use this information to find projects that are not under the same folder.  Currently the only used one is `wrap` so `tests` is kind of useless.

**Comment by jchannon** [comment]
We have a dependency on DependcyModel in Nancy that'd we'd like to get rid
of ideally as this also forces us on netstandard16

**Comment by jeremymeng** [comment]
> Why do we have to pass the packages folder into the msbuild command? Are we using a non-standard package folder location?

IIRC NuGet restore uses `.\packages` folder for `packages.config` based project, and `%USERPROFILE%\.nuget\packges` for `project.json` based project.  Because the appveyor build machine has limited storage space under `%USERPROFILE%`  I passed the option to make it the same `.\packages` folder.

**Comment by jeremymeng** [comment]
> .nuget.targets in .gitignore seems redundant. I couldn't find a way to get such a file to appear.

It could be from Oren's [Nuspec.ReferenceGenerator](https://github.com/onovotny/ReferenceGenerator), which automatically manages the dependency groups in .nuspec files.  However it is not used currently so it is indeed redundant.

**Comment by adamralph** [comment]
Thanks for the responses @jeremymeng.

I've added a checklist item to remove the redundant project section. (It was only required temporarily by @jeremymeng while he had the `net45` project unloaded from the solution.)

Regarding `dotnet restore` using `global.json`, we also have a `project.json` under `tests\FakeItEasy.Tests.dotnet` so I would have thought that it *is* required to have `tests` in this file. Regardless, if it isn't hurting anything, I think I'd like to include all folders in this file so that when we next add a project which needs `dotnet restore`, it will automatically be catered for.

I've added a checklist item to determine if we can remove the dep on `DependencyModel`.

**Comment by adamralph** [comment]
FYI we (@FakeItEasy/owners) are considering relaxing our policy on major version bumps and releasing 3.0 before we release netstd support, while acknowledging that unforeseen changes may require us to bump the major version again to 4.0 when we release netstd support. We have yet to reach consensus on this and the discussion is on-going.

**Comment by adamralph** [comment]
@jeremymeng's regarding [your comment about the version warnings](https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-237001541) - AFAIK this was fixed by @blairconrad in #811.

**Comment by adamralph** [comment]
> - [ ] global.json should be added to "Solution Items" in FakeItEasy.sln, even if it's not relevant in that context (IMHO all files in the root folder should be added).

I've redacted this checklist item. We're already not including irrelevant content in FakeItEasy.sln, e.g. the folders and files for the `dotnet` projects.

**Comment by adamralph** [comment]
The [checklist](https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-236366247) is dwindling!

The item that interests me most right now is this one:

> Why are we using csproj for FakeItEasy.dotnet and xproj for FakeItEasy.Tests.dotnet?

**Comment by adamralph** [comment]
Regarding this item:

> Is `dotnet` the best moniker to distinguish the new bits? Since the aim is to support the `netstandard1.6` TFN I would have thought `netstandard` would be more appropriate. (Or `netstd`.)

Given that `dotnet` is the name of the exe that does all the magic in the new world, perhaps `dotnet` is the best moniker?

On that note, should our [new rakefile](https://github.com/FakeItEasy/FakeItEasy/pull/815) be using `dotnet` to build the new projects instead of msbuild.exe?

**Comment by thomaslevesque** [comment]
I think I'd prefer `netstd` or `netcore`. I find it weird to call it `dotnet`, since the existing solution also targets .NET...

**Comment by adamralph** [comment]
If we're not going with `dotnet` then I would go with `netstd`. .NET Core is just one of the platforms we're going to be supporting via our new .NET Platform Standard assembly.

**Comment by jeremymeng** [comment]
Re: csproj vs xproj

I chose project.json/xproj mainly because I put the new .NET Core project in a new folder, and csproj has better VS IDE support when source files are in a different folder.

The test project has to be xproj because it's the only supported way to use `dotnet test`.

For more pros and cons of the two project flavors,  see https://docs.microsoft.com/en-us/dotnet/articles/core/porting/project-structure

Sent from Outlook Mobile<https://aka.ms/blhgte>.

**Comment by adamralph** [comment]
Thanks @jeremymeng. I'm quite happy with `FakeItEasy-dotnet` being a `csproj`. For one thing, it allows [installation of `targets` nuget packages](https://github.com/FakeItEasy/FakeItEasy/pull/817/files#r73284144).

> The test project has to be xproj because it's the only supported way to use `dotnet test`.

Judging by the `project.json`, `FakeItEasy.Test-dotnet` is a `netcoreapp1.0` project. Is `dotnet test` the only way to invoke xunit tests against such an animal? Or is there another species of project we could use to test the netstd library which has another method?

(I'm still quite confused by it all.)

**Comment by adamralph** [comment]
I see no mention of the approach we have taken (build a netstd library using csproj) at https://docs.microsoft.com/en-us/dotnet/articles/core/porting/project-structure.

**Comment by jeremymeng** [comment]
It is the alternative under PCL. The article doesn't mention test project or might have incorrectly used csproj for test project for netstandard PCL.

Sent from Outlook Mobile<https://aka.ms/blhgte>.

**Comment by adamralph** [comment]
I see, so we're using "Create a Portable Class Library (PCL) to target .NET Core" and "Alternative: Create the PCL in another folder and reference the original source code to avoid this issue.".

And yes, it looks like the article is saying that the test project for the PCL should also be csproj. Why is that not possible?

**Comment by blairconrad** [comment]
> Why are we using csproj for FakeItEasy.dotnet and xproj for FakeItEasy.Tests.dotnet?

For fun, I tried copying the specs to a .NET Standard-targeting csproj, but failed miserably. Couldn't reference FakeItEasy.Tests.stdnet (in my [branch](https://github.com/FakeItEasy/FakeItEasy/pull/821), but you know it as .dotnet) from the new Specs projects because the former is a .xproj. I think .xprojness is viral. But up 'til then, things seemed to go well.

**Comment by jeremymeng** [comment]
> Regarding dotnet restore using global.json, we also have a project.json under tests\FakeItEasy.Tests.dotnet so I would have thought that it is required to have tests in this file.

The list of folders in global.json are searched to find project references during `dotnet restore` when the references cannot be found under the same parent folder. They are not folders to run restore command. Restore is run for each project.json found recursively.

**Comment by jeremymeng** [comment]
> And yes, it looks like the article is saying that the test project for the PCL should also be csproj. Why is that not possible?

The only official way that I know of for testing .NET Core is `dotnet test` in a .xproj/project.json project.  I am working with the doc author to have the section updated.

**Comment by adamralph** [comment]
> The list of folders in global.json are searched to find project references during dotnet restore when the references cannot be found under the same parent folder. They are not folders to run restore command. Restore is run for each project.json found recursively.

Something's still not clicking for me here. I'm not sure that I understand anything in this paragraph.

**Comment by adamralph** [comment]
> The only official way that I know of for testing .NET Core is dotnet test in a .xproj/project.json project. I am working with the doc author to have the section updated.

I guess this begs the question: "When producing a netstandard project, on what supporting platforms should run my tests for the project?" It could be just one of them, all of them, or some subset. In our case, our packaged assembly is `netstandard1.6`, which is supported by `netcoreapp1.0` and `net463` (not yet released). We are only testing on one of those supporting platforms, `netcoreapp1.0`, but we could equally test on `net463` instead (if it were released), which would be supported by both xunit and xbehave through backwards compatibility with `net45`. Until `net463` is released, `netcoreapp1.0` is our only option so we'll have to go with the current `.xproj/project.json` project for now.

**Comment by blairconrad** [comment]
>  Determine if we can remove the dependency on https://www.nuget.org/packages/Microsoft.Extensions.DependencyModel/ before RTM and therefore target netstd 1.3 instead of netstd 1.6

DependencyModel gives us the ability to investigate the assemblies that the tests depend on. The best discussion of this that I've found is at https://github.com/dotnet/coreclr/issues/919. Maybe someone knows better.

Anyhow, we walk the assemblies to find all available types, which lets us find the Bootstrapper to use as well as any user-defined Dummy Factories, Fake Object Builders, and Argument Value Formatters.

An option could be to release a version of the library that targets 1.3, if users didn't mind being restricted to the built-in Dummy Factories etc. Then they could use the 1.6 flavour for full power.

As an alternative, we could look at having another way to tell FakeItEasy where to find these classes, such as with an explicit call to some static FakeItEasy method.
The difficulty is that then we'd have to rely on a facility to run the setup code before any tests are run.
Or I suppose rejig the "extension point" code to support just-in-time discovery of the Factories, Builders, and Formatters. But that could be inconvenient, as users still may have to litter their tests with directives to tell FakeItEasy what to use, just to make sure that requirement has been met.

**Comment by blairconrad** [comment]
Oh, I should've said. The "release a 1.3 and 1.6 .NET Standard" option is not something I thought of. I see that StructureMap has done a similar thing.

**Comment by adamralph** [comment]
Of the options presented above, I'd probably either go with just supporting netstd1.6, or supporting both netstd1.3 and netstd1.6, with the former not supporting those user defined bits at all. I.e. I don't think it's worth introducing the complexity of specialised extension points just for netstd1.3.

**Comment by blairconrad** [comment]
Oh, were we to introduce a different method of finding the extension points, I'd've thought to make it the single way, for all frameworks.

**Comment by adamralph** [comment]
Fair enough, but I don't think it's worth changing that for all frameworks, just to support netstd1.3. Moreover, I don't think it's worth doing *anything* non-trivial to support netstd1.3 over 1.6. The only thing that 1.3 gives us over 1.6 is support for Universal Windows Platform 10.0.

**Comment by blairconrad** [comment]
> Fair enough, but I don't think it's worth changing that for all frameworks, just to support netstd1.3. 

No, probably not. Although the current solution is sometimes cumbersome in its own right. Then again, we've at least not had any complaints around its performance for some time. 

> I don't think it's worth doing anything non-trivial to support netstd1.3 over 1.6. The only thing that 1.3 gives us over 1.6 is support for Universal Windows Platform 10.0.

Agreed. Does this check the box in the list, then?

**Comment by adamralph** [comment]
Actually, targetting netstd <1.6 does give us one distinct advantage. It allows us to test the lib using a net462 project rather than a netcoreapp1.0 project. That will allow us to fill in all the missing gaps around testing right now.

As soon as .NET 4.6.3 is released, this advantage vanishes, but I don't know what the ETA is on that.

There is also @jeremymeng's [comment](https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-237004299) above:

> I also heard that this code (DependencyModel) might be going away in the future. I don't know more details yet.

Before checking off that item, I'd like to know whether this is true, and if so, when (and how) it will "go away".

**Comment by khellang** [comment]
Won't you penalize users on full framework by only targeting netstandard? You're basically forcing them to pull in NuGet packages for dependencies they might already have on disk.

**Comment by adamralph** [comment]
In an effort to make [The List](https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-236366247) the complete list of things to do before we can close this issue, I've added:

> - [ ] We don't have `build.cmd` and `rakefile` equivalents for producing a package containing both net40 and netstd assemblies
> - [ ] We don't have CI which produces a package containing both net40 and netstd assemblies

**Comment by adamralph** [comment]
@khellang we won't only support netstd. We'll continue to support net40 explicitly too.

**Comment by khellang** [comment]
> @khellang we won't only support netstd. We'll continue to support net40 explicitly too.

Ah. I just didn't see it in the project.json file.

**Comment by adamralph** [comment]
@khellang it's a separate project.

Having said that, we have previously discussed a distant future where, once we decide to drop support for older frameworks, we do indeed ship just a single netstd assembly. I would have thought that ultimately that would be the goal for any nuget package?

**Comment by khellang** [comment]
> @khellang it's a separate project.

Why keep a separate project when project.json does such an awesome job at cross-compiling and packaging?

> I would have thought that ultimately that would be the goal for any nuget package?

Once Microsoft stops supporting the full, desktop .NET framework, sure. I don't think that'll happen very soon, though. So until then, I'd prefer libs to keep `net*` targets around.

**Comment by jeremymeng** [comment]
>> The list of folders in global.json are searched to find project references during dotnet restore when the references cannot be found under the same parent folder. They are not folders to run restore command. Restore is run for each project.json found recursively.

> Something's still not clicking for me here. I'm not sure that I understand anything in this paragraph.

Let me try with examples :)

* when you run `dotnet restore` in a folder, `dotnet` searches for `project.json` files recursively, and run restore on each of them regardless of what's specified in `global.json`.
* when you have `src\A` and `src\B`, and `B` references `A`, you don't need `src` in `global.json` since `A` and `B` are under the same folder.
* when you have `src\lib` and `tests\libtests`, and `libtests` references `lib` project, restoring for `tests\libtests` would fail to find `lib`, unless you add `src` in `global.json` to tell `dotnet` where to look for referenced projects.

**Comment by jeremymeng** [comment]
>> I also heard that this code (DependencyModel) might be going away in the future. I don't know more details yet.

> Before checking off that item, I'd like to know whether this is true, and if so, when (and how) it will "go away".

Actually my statement is incorrect saying it might be going away.  I am not sure what would happen when we move to MSBuild.

**Comment by blairconrad** [comment]
> Why keep a separate project when project.json does such an awesome job at cross-compiling and packaging?

We're not yet ready to force contributors to have the .NET Core tooling.

**Comment by jeremymeng** [comment]
> but we could equally test on net463 instead (if it were released)

Agree on that. I think one platform to test the netstandard code path is enough.  One thing to note is that by default for class library projects, nuget package references are not copied to the output directory so tests would fail to load indirect nuget package dependencies. Setting `CopyNuGetImplementations` to true overrides the behavior. I've verified that the following modifications are needed for a net46 test project to test a netstandard1.3 PCL project:

1.	Manually edit the test .csproj file to add `<CopyNuGetImplementations>true</CopyNuGetImplementations>`
1.	Remove `packages.config` and add a `project.json` to list nuget dependencies
1.	Add `"runtimes": { "win": {} }` to the added `project.json` of the test project

**Comment by adamralph** [comment]
@jeremymeng thanks for the explanation of global.json.

The thing I still find odd is:

> when you have src\lib and tests\libtests, and libtests references lib project, restoring for tests\libtests would fail to find lib, unless you add src in global.json to tell dotnet where to look for referenced projects.

Since "libtests references lib project", I still don't understand why global.json is required. Why can't the tooling just follow the reference?

**Comment by khellang** [comment]
> Since "libtests references lib project", I still don't understand why global.json is required. Why can't the tooling just follow the reference?

@adamralph Because the reference could be a NuGet package from NuGet.org.

**Comment by khellang** [comment]
"Let me first check global.json... Oh, I have a local project called "lib", let me use that instead of going to NuGet.org."

**Comment by adamralph** [comment]
@khellang I see, thanks. I'm still a project.json noob. So there's no difference between project and package references?

**Comment by khellang** [comment]
> So there's no difference between project and package references?

That's right. EVERYTHING is NuGet packages now. No assembly or project references.

**Comment by adamralph** [comment]
Thanks @khellang. I wonder if a better name could have been chosen than "projects" for that property to make this more intuitive.

In that case, it seems we don't need `samples` in global.json, so I'll check off that item from the list.

**Comment by adamralph** [comment]
I've checked off

> Determine if we can remove the dependency on https://www.nuget.org/packages/Microsoft.Extensions.DependencyModel/ before RTM and therefore target netstd 1.3 instead of netstd 1.6

I think it's best to assume it's going nowhere for now, which means we'll have to continue using `netcoreapp` (.xproj+project.json) projects to test the netstd1.6 assembly.

If, in the meantime, .NET Framework 4.6.3 is released, we can consider switching to that to make things easier, but I think we should also assume that that isn't going to happen before we release our netstd1.6 assembly.

**Comment by jeremymeng** [comment]
> Since "libtests references lib project", I still don't understand why global.json is required. Why can't the tooling just follow the reference?

The tooling might not know where to find the `lib` project.  For example, when you run `dotnet restore` command under `libtests\` directory, how would the tooling probe outside of this directory?  Another option is that the tooling could go up until it finds a `global.json`, then use that directory as the search root.  Anyway the current design is https://github.com/NuGet/NuGet.Client/blob/125b3bbd41dec25abfae8dd8d64627f78aae41d0/src/NuGet.Core/NuGet.ProjectModel/PackageSpecResolver.cs#L93-L97

**Comment by adamralph** [comment]
Three more items checked off [The List](https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-236366247) in #821:

> - [x] Is dotnet the best moniker to distinguish the new bits? Since the aim is to support the netstandard1.6 TFN I would have thought netstandard would be more appropriate. (Or netstd.)
> - [x] We don't have build.cmd and rakefile equivalents for producing a package containing both net40 and netstd assemblies (@blairconrad: #821 works toward this)
> - [x] We don't have CI which produces a package containing both net40 and netstd assemblies (@blairconrad: #821 works toward this, aiming to use Appveyor)

I think some of the naming (`netstd` vs `dotnet` vs `netcoreapp`) still isn't optimal, but it can be dealt with later, and may even be somewhat opinionated.

**Comment by enzian** [comment]
I've been using the beta package and so far everything seems fine! Thanks a lot for porting - I really appreciate the effort you guys put in!

**Comment by blairconrad** [comment]
@enzian, that's really nice to hear. Thanks for the words of encouragement!

**Comment by adamralph** [comment]
How are we affected by the recent revelation that netstandard 1.5 and 1.6 are going to be broken by 2.0?

**Comment by khellang** [comment]
I'm not sure the breaking change is real after all; https://twitter.com/terrajobst/status/786993233913978880

**Comment by adamralph** [comment]
Thanks @khellang, I missed that. Good to know!

**Comment by adamralph** [comment]
Assigned to 3.0.0 milestone.

**Comment by adamralph** [comment]
I checked off the integration tests item in [The List](https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-236366247).

Only specs to go! That's a good incentive to sort out xbehave once and for all.

BTW - I'd be content to push an 3.0 alpha to nuget.org now if we want to unblock https://github.com/NancyFx/Nancy/issues/2612 and then push the first beta once we've got the specs running.

**Comment by thomaslevesque** [comment]
> I'd be content to push an 3.0 alpha to nuget.org now if we want to unblock NancyFx/Nancy#2612 and then push the first beta once we've got the specs running.

Sounds like a good plan 👍

**Comment by blairconrad** [comment]
Oh, bother. I'd made a beta001. Can make an alpha001.

**Comment by adamralph** [comment]
FYI I just pushed a xbehave 2.2.0 beta 1 to NuGet, with support for .NET Standard 1.0! https://www.nuget.org/packages/Xbehave/2.2.0-beta0001-build680

**Comment by thomaslevesque** [comment]
> I'll expect your "hook up the specs" PR when I wake up!

I'm working on making the specs work on netcoreapp1.0 now. I haven't done much for .NET Core support so far, so I guess it's my last chance :wink:

**Comment by blairconrad** [comment]
Haha! I was just having some fun.

Also, welcome to heck.

**Comment by adamralph** [comment]
The project.json for the test project in xbehave itself might be a good
reference. The explicit use of the testutility package should be ignored
though, that's for the meta level code which allows scenarios to test
scenarios.

**Comment by thomaslevesque** [comment]
> Also, welcome to Heck.

Well, the worst part is the poor VS integration. No Intellisense, no test runner and no debug for netstd 😒 

> The project.json for the test project in xbehave itself might be a good reference. The explicit use of the testutility package should be ignored though, that's for the meta level code which allows scenarios to test scenarios.

Thanks; actually, I copied and edited the project.json from integration tests, and it's working fine. I just have one failing scenario to fix.

**Comment by adamralph** [comment]
The VS test runner should work. It works for the xbehave solution.

On 28 Nov 2016 11:13 p.m., "Thomas Levesque" <notifications@github.com>
wrote:

Also, welcome to Heck.

Well, the worst part is the poor VS integration. No Intellisense, no test
runner and no debug for netstd 😒

The project.json for the test project in xbehave itself might be a good
reference. The explicit use of the testutility package should be ignored
though, that's for the meta level code which allows scenarios to test
scenarios.

Thanks; actually, I copied and edited the project.json from integration
tests, and it's working fine. I just have one failing scenario to fix.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-263411279>,
or mute the thread

**Comment by thomaslevesque** [comment]
> The VS test runner should work. It works for the xbehave solution.

For me it doesn't see *any* test, not even the net40 ones. Maybe I need to install something else.

**Comment by adamralph** [comment]
[The List](https://github.com/FakeItEasy/FakeItEasy/issues/531#issuecomment-236366247) is complete!

Fantastic work all round.

**Comment by blairconrad** [comment]
Huzzah! Well done, everyone!

![image](https://cloud.githubusercontent.com/assets/3275797/20748218/4df01d5e-b6bb-11e6-91c4-14d639f362f0.png)

**Comment by adamralph** [comment]
![](http://static.celebuzz.com/uploads/2013/06/14/request-five.gif)

**Comment by blairconrad** [comment]
This issue has been released in FakeItEasy 3.0.0:
https://github.com/FakeItEasy/FakeItEasy/releases/3.0.0

@jeremymeng, @jonorossi, thanks for all your help.
Look for your names in the release notes! :trophy:

**Comment by weitzhandler** [comment]
Can you please add support for modern Xamarin PCLs (Profile 44), that target only UWP, Droid and iOS?

**Comment by blairconrad** [comment]
@weitzhandler, thanks for your interest in FakeItEasy. I'll be the first to admit that I don't understand all the PCLs and whatnot, but I don't think we'll do this.
For starters, FakeItEasy relies on Castle.Core, which would have to support the profile, and it supports only .NETFramework 3.5, .NETFramework 4.0, .NETFramework 4.5, and .NETStandard 1.3. Also, I think the PCLs are the wave of the past, and we're looking forward to the [.NET Standard](http://adamralph.com/netstandard-deck/).

My understanding is that the usual course would be to build an app's business logic into a PCL such that supports your target platform as well as .NET 4.0 or 4.5 or similar and then to have the test assemblies target that .NET Framework (or a compatible one). Profile44 supports .NET 4.5.1, so this should be an option for you.
